### PR TITLE
ICU-21569 LSTM Optimize memory usage.

### PR DESCRIPTION
**Pull request by FrankYFTang**
Refactor the array allocation out of compute method and out of the loop
to reduce repeated malloc/free. Move all the array allocation code to
the same place before the loop.
This should dramatically reduce the number of calls to malloc/free.

##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21569
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by roubert** [comment]
Hm, this is all code I've never seen before. Is there maybe someone who has seen it before who could review this instead?

**Comment by FrankYFTang** [comment]
> Hm, this is all code I've never seen before. Is there maybe someone who has seen it before who could review this instead?

I recently added this code. This PR is a simple mechanical refactoring only. Basically we used to have

for loop {
   call compute(arg1a, arg2a ... arg xa, status)
}
...
for loop {
   call compute(arg1b, arg2b ... arg xb, status)
}
compute () {
  Array1D hU
  // use hU as temp memory
....
  Array1D ic
  // use ic as temp memory
...
}

and I now change it to 

  Array1D hU
  Array1D ic
for loop {
   call compute(arg1a, arg2a ... arg xa, w1, w2)
}
...
for loop {
   call compute(arg1b, arg2b ... arg xb, w1, w2)
}
compute ( .... Array1D& hu, Array1D& ic) {
  // use hU as temp memory
....
  // use ic as temp memory
...
}

so hU and ic will only be called allocated on the stack once and won't need to be called inside the compute repeatedly

Notice, there are two version of Array1D constructor, the one which passed in a status will malloc and the other just take the memory which do not need to malloc . The one which take the pass in memory is cheap to allocate on the stack, the one which take a number and status which malloc internally is expensive. 

I change move all the calls which need to malloc inside compute out so we no longer need to pass in status to the compute and check after the return. All the status check of memory allocation now is done once before the first for loop.

There are really no logic change in this PR, just moving the code out of function and move some line up in the function. All the code before/after should be logical equivalent

File: icu4c/source/common/lstmbe.cpp [hunk 1]
```diff
@@ -589,36 +589,29 @@ void GraphemeClusterVectorizer::vectorize(
...
-    Array1D ifco(b.d1(), status);
-    {
-        Array1D hU(b.d1(), status);
-        if (U_FAILURE(status)) return;
-        ifco.dotProduct(x, W)
-            .add(hU.dotProduct(h, U))
-            .add(b);
-    } // delocate hU
+    ifco.dotProduct(x, W)
+        .add(hU.dotProduct(h, U))
```
Comment (sffc) [inline icu4c/source/common/lstmbe.cpp:603]: Suggestion (optional): Add a new specialized function `.addDotProduct` that avoids the need for allocating `hU`

**Comment by FrankYFTang** [comment]
PTAL

**Review by sffc — DISMISSED** [review]

**Comment by FrankYFTang** [comment]
SQUASHED, please approve again.

**Review by sffc — APPROVED** [review]

**Pull request merged**